add post (#190)
rstudio/ai-blog: _posts/2022-10-06-audio-classification-torch/audio-classification.Rmd

---
title: "Audio classification with torch"
description: >

>   Learn how to classify speech utterances with torch, making use of domain knowledge and deep learning. This post is a condensed version of the same-topic chapter in the book to be published with CRC Press, Deep Learning and Scientific Computing with R torch. 

Reviewer: ```suggestion
  Learn how to classify speech utterances with torch, making use of domain knowledge and deep learning. This post is a condensed version of the corresponding chapter in the forthcoming book, Deep Learning and Scientific Computing with R torch, to be published by CRC Press. 
```

Reviewer: Tried to re-jig the words to frontload the most important info (like the name of the book, versus the name of the publisher).

Author: thanks! taking it as a compliment for my politeness ;-))))

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanatorchaudio
date: 2022-10-06
categories:

>   - Torch

Reviewer: I haven't checked, but is Torch always capitalized in the tags? I'm inferring that it's capitalized when referring to Torch (the library) versus small-t torch the R package (bonus confusion for the fact that the native package is also small-t it looks like from [the Wikipedia page](https://en.wikipedia.org/wiki/Torch_(machine_learning))).

I think it's fine either way—it's obvious from the fact that both Torch and R are in the tags that you're going to discuss the R package for it—just checking for consistency.

Author: it's just the category really that is capitalized (for aesthetical reasons, because it would look weird otherwise)

  - R
  - Audio Processing
bibliography: bibliography.bib
output:
  distill::distill_article:
    self_contained: false
    toc: true
preview: images/squirrel.jpg
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
```

## Variations on a theme

> [Simple Audio Classification with Keras](https://blogs.rstudio.com/ai/posts/2018-06-06-simple-audio-classification-keras/), [Audio classification with Keras: Looking closer at the non-deep learning parts](https://blogs.rstudio.com/ai/posts/2019-02-07-audio-background/), [Simple audio classification with torch](https://blogs.rstudio.com/ai/posts/2021-02-04-simple-audio-classification-with-torch/): No, this is not the first post, on this blog, that introduces speech classification using deep learning. With two of those posts (the "applied" ones) it shares the general setup, the type of deep-learning architecture employed, and the dataset used. With the third, it has in common the interest in the ideas and concepts involved. Each of these posts has a different focus, -- should you read this one?

Reviewer: It looks like the three of you actually do use different capitalization rules for your post titles. But, since you and Athos are in the majority, I think Daniel can get overriden here for consistency's sake.

Author: :-))

> [Simple Audio Classification with Keras](https://blogs.rstudio.com/ai/posts/2018-06-06-simple-audio-classification-keras/), [Audio classification with Keras: Looking closer at the non-deep learning parts](https://blogs.rstudio.com/ai/posts/2019-02-07-audio-background/), [Simple audio classification with torch](https://blogs.rstudio.com/ai/posts/2021-02-04-simple-audio-classification-with-torch/): No, this is not the first post, on this blog, that introduces speech classification using deep learning. With two of those posts (the "applied" ones) it shares the general setup, the type of deep-learning architecture employed, and the dataset used. With the third, it has in common the interest in the ideas and concepts involved. Each of these posts has a different focus, -- should you read this one?

Reviewer: ```suggestion
[Simple audio classification with Keras](https://blogs.rstudio.com/ai/posts/2018-06-06-simple-audio-classification-keras/), [Audio classification with Keras: Looking closer at the non-deep learning parts](https://blogs.rstudio.com/ai/posts/2019-02-07-audio-background/), [Simple audio classification with torch](https://blogs.rstudio.com/ai/posts/2021-02-04-simple-audio-classification-with-torch/): No, this is not the first post on this blog that introduces speech classification using deep learning. With two of those posts (the "applied" ones) it shares the general setup, the type of deep-learning architecture employed, and the dataset used. With the third, it has in common the interest in the ideas and concepts involved. Each of these posts has a different focus -- should you read this one?
```

> Well, of course I can't say "no" here -- all the more that here, you have an abbreviated and condensed version of the same-topic chapter that is part of the book to be published by CRC Press, *Deep Learning and Scientific Computing with R `torch`*. By way of comparison with the previous post that used `torch`, written by the creator and maintainer of `torchaudio`, Athos Damiani, significant developments have taken place in the `torch` ecosystem, the end result being that the code got a lot easier (especially in the model training part). That said, let's end the preamble already, and plunge into the topic!

Reviewer: ```suggestion
Well, of course I can't say "no" -- all the more so because, here, you have an abbreviated and condensed version of the chapter on this topic in the forthcoming book from CRC Press, *Deep Learning and Scientific Computing with R `torch`*. By way of comparison with the previous post that used `torch`, written by the creator and maintainer of `torchaudio`, Athos Damiani, significant developments have taken place in the `torch` ecosystem, the end result being that the code got a lot easier (especially in the model training part). That said, let's end the preamble already, and plunge into the topic!
```

Reviewer: Tried to remove the double use of "here", and re-work things to avoid "same-topic chapter" (which sounds like a strange phrase to my ear).

Author: > all the more so because,
oh right, that was what I wanted to say, but I didn't get it right :-)

> from CRC Press
prepositions are hard, in English (well, in any language, I guess)

## Inspecting the data

> We use the *speech commands* dataset (@abs-1804-03209) that comes with [`torchaudio`](https://github.com/curso-r/torchaudio). The dataset holds recordings of thirty different, one- or two-syllable words, uttered by different speakers; there are about 65,000 audio files overall. Our task will be to predict, from the audio solely, which of thirty possible words was pronounced.

Reviewer: ```suggestion
We use the *speech commands* dataset (@abs-1804-03209) that comes with [`torchaudio`](https://github.com/curso-r/torchaudio). The dataset holds recordings of thirty different one- or two-syllable words, uttered by different speakers. There are about 65,000 audio files overall. Our task will be to predict, from the audio solely, which of thirty possible words was pronounced.
```

Reviewer: You know me, always out to destroy semi-colons 😈😉

Author: hehe :-)
Of course, I have mixed thoughts on this ;-)))
Mixed, because in the book, I've taken care to keep sentences short (or at least, that was the intention)

```{r}
sample$sample_rate
```

    [1] 16000

All recordings have been sampled at the same rate. Their length almost always equals one second; the -- very -- few sounds that are minimally longer we can safely truncate.

Finally, the target is stored, in integer form, in `sample$label_index`, the corresponding word being available from `sample$label`:

```{r}
sample$label
sample$label_index
```

    [1] "bird"
    torch_tensor
    2
    [ CPULongType{} ]

How does this audio signal "look"?

```{r}
library(ggplot2)

df <- data.frame(
  x = 1:length(sample$waveform[1]),
  y = as.numeric(sample$waveform[1])
  )

ggplot(df, aes(x = x, y = y)) +
  geom_line(size = 0.3) +
  ggtitle(
    paste0(
      "The spoken word \"", sample$label, "\": Sound wave"
    )
  ) +
  xlab("time") +
  ylab("amplitude") +
  theme_minimal()
```

![The spoken word "bird", in time-domain representation.](images/audio-bird-waveform.png){#fig-audio-bird-waveform fig-alt="A sound wave, displaying amplitude over time."}

> What we see is a sequence of amplitudes, reflecting the sound wave produced by someone saying "bird". Put differently, we have here a time series of "loudness values". Even for experts, guessing *which* word resulted in those amplitudes is an impossible task. This is where domain knowledge comes in. The expert may not be able to make much of the signal *in this representation*; but they may know a way to more meaningfully represent it.

Reviewer: Really great paragraph!

Author: thanks!!

## Two equivalent representations

> Imagine that instead of as a sequence of amplitudes over time, the above wave were represented in a way that had no information about time at all. Next, imagine we took that representation and tried to recover the original signal. For that to be possible, the new representation would somehow have to contain "just as much" information as the wave we started from. That "just as much" is obtained from the *Fourier Transform*, and it consists of the magnitudes and phase shifts of the different *frequencies* that make up the signal.

Reviewer: > the above wave were represented

I'm not sure if you want plural ("waves were") or singular ("wave was"), but the noun and verb should agree, either way.

Otherwise, really nice explanation.

Author: it's supposed to be conjunctive (I like the _were_ ;-))

How, then, does the Fourier-transformed version of the "bird" sound wave look? We obtain it by calling `torch_fft_fft()` (where `fft` stands for Fast Fourier Transform):

```{r}
dft <- torch_fft_fft(sample$waveform)
dim(dft)
```

    [1]     1 16000

The length of this tensor is the same; however, its values are not in chronological order. Instead, they represent the *Fourier coefficients*, corresponding to the frequencies contained in the signal. The higher their magnitude, the more they contribute to the signal:

```{r}
mag <- torch_abs(dft[1, ])

df <- data.frame(
  x = 1:(length(sample$waveform[1]) / 2),
  y = as.numeric(mag[1:8000])
)

ggplot(df, aes(x = x, y = y)) +
  geom_line(size = 0.3) +
  ggtitle(
    paste0(
      "The spoken word \"",
      sample$label,
      "\": Discrete Fourier Transform"
    )
  ) +
  xlab("frequency") +
  ylab("magnitude") +
  theme_minimal()
```

![The spoken word "bird", in frequency-domain representation.](images/audio-bird-dft.png){#fig-audio-bird-dft fig-alt="A curve showing magnitude for Fourier bins from 0 to 8000. Above 4000, nearly all are zero."}

> From this, alternate, representation, we could go back to the original sound wave by taking the frequencies present in the signal, weighting them according to their coefficients, and adding them up. But in sound classification, timing information must surely matter; we don't really want to throw it away.

Reviewer: ```suggestion
From this alternate representation, we could go back to the original sound wave by taking the frequencies present in the signal, weighting them according to their coefficients, and adding them up. But in sound classification, timing information must surely matter; we don't really want to throw it away.
```

## Combining representations: The spectrogram

In fact, what really would help us is a synthesis of both representations; some sort of "have your cake and eat it, too". What if we could divide the signal into small chunks, and run the Fourier Transform on each of them? As you may have guessed from this lead-up, this indeed is something we can do; and the representation it creates is called the *spectrogram*.

> With a spectrogram, we still keep some time-domain information -- some, since there is an unavoidable loss in granularity. On the other hand, for each of the time segments, we learn about their spectral composition. There's an important point to be made, though. The resolutions we get in *time* versus in *frequency*, respectively, are inversely related. If we split up the signals into many chunks (called "windows"), the frequency representation per window will not be very fine-grained. Conversely, if we want to get better resolution in the frequency domain, we have to choose longer windows, thus losing information about how spectral composition varies over time. What sounds like a big problem -- and in many cases, will be -- won't be one for us, though, as you'll see very soon.

Reviewer: 👍

```{r}
fft_size <- 512
window_size <- 512
power <- 0.5

spectrogram <- transform_spectrogram(
  n_fft = fft_size,
  win_length = window_size,
  normalized = TRUE,
  power = power
)

spec <- spectrogram(sample$waveform)$squeeze()
dim(spec)

```

    [1]   257 63

We can display the spectrogram visually:

```{r}
bins <- 1:dim(spec)[1]
freqs <- bins / (fft_size / 2 + 1) * sample$sample_rate 
log_freqs <- log10(freqs)

frames <- 1:(dim(spec)[2])
seconds <- (frames / dim(spec)[2]) *
  (dim(sample$waveform$squeeze())[1] / sample$sample_rate)

image(x = as.numeric(seconds),
      y = log_freqs,
      z = t(as.matrix(spec)),
      ylab = 'log frequency [Hz]',
      xlab = 'time [s]',
      col = hcl.colors(12, palette = "viridis")
)
main <- paste0("Spectrogram, window size = ", window_size)
sub <- "Magnitude (square root)"
mtext(side = 3, line = 2, at = 0, adj = 0, cex = 1.3, main)
mtext(side = 3, line = 1, at = 0, adj = 0, cex = 1, sub)

```

![The spoken word "bird": Spectrogram.](images/audio-spectrogram.png){#fig-audio-spectrogram fig-alt="A three-dimensional plot where the x-axis represents time, the y-axis, frequency, and color, magnitude."}

> We know that we've lost some resolution, in both time and frequency. By displaying the square root of the coefficients' magnitudes, though -- and thus, enhancing sensitivity -- we were still able to obtain a reasonable result. (With the `viridis` color scheme, long-wave shades indicate higher-valued coefficients; short-wave ones, the opposite.)

Reviewer: ```suggestion
We know that we've lost some resolution in both time and frequency. By displaying the square root of the coefficients' magnitudes, though -- and thus, enhancing sensitivity -- we were still able to obtain a reasonable result. (With the `viridis` color scheme, long-wave shades indicate higher-valued coefficients; short-wave ones, the opposite.)
```

    list(x = x, y = y)
  }
)

```

In the parameter list to `spectrogram_dataset()`, note `power`, with a default value of 2. This is the value that, unless told otherwise, `torch`'s `transform_spectrogram()` will assume that `power` should have. Under these circumstances, the values that make up the spectrogram are the squared magnitudes of the Fourier coefficients. Using `power`, you can change the default, and specify, for example, that'd you'd like absolute values (`power = 1`), any other positive value (such as `0.5`, the one we used above to display a concrete example) -- or both the real and imaginary parts of the coefficients (`power = NULL)`.

Display-wise, of course, the full complex representation is inconvenient; the spectrogram plot would need an additional dimension. But we may well wonder whether a neural network could profit from the additional information contained in the "whole" complex number. After all, when reducing to magnitudes we lose the phase shifts for the individual coefficients, which might contain usable information. In fact, my tests showed that it did; use of the complex values resulted in enhanced classification accuracy. [^1]

> [^1]: In the book, I'm comparing three different setups (the third being use of Mel spectrograms). Using complex coefficients for network input was the winner.

Reviewer: ```suggestion
[^1]: In the book, I compare three different setups (the third being use of Mel spectrograms). Using complex coefficients for network input was the winner.
```

Let's see what we get from `spectrogram_dataset()`:

```{r}
ds <- spectrogram_dataset(
  root = "~/.torch-datasets",
  url = "speech_commands_v0.01",
  download = TRUE,
  power = NULL
)

dim(ds[1]$x)
```

    [1]   2 257 101

> We have two hundred fifty-seven coefficients for a hundred and one windows; and each coefficient is represented by both its real and imaginary parts.

Reviewer: ```suggestion
We have 257 coefficients for 101 windows; and each coefficient is represented by both its real and imaginary parts.
```

Reviewer: Numbers over ten (or twenty, depending on whom you ask) are written in Arabic numerals. ✅

Author: oh, see, so I over-generalized from some correction you made, in the past :-)

Let's check actual accuracies.

    "epoch","set","loss","acc"
    1,"train",3.09768574611813,0.12396992171405
    1,"valid",2.52993751740923,0.284378862793572
    2,"train",2.26747255972008,0.333642356819118
    2,"valid",1.66693911248562,0.540791100123609
    3,"train",1.62294889937818,0.518464153275649
    3,"valid",1.11740599192825,0.704882571075402
    ...
    ...
    38,"train",0.18717994078312,0.943809229501442
    38,"valid",0.23587799138006,0.936418417799753
    39,"train",0.19338578602993,0.942882159044087
    39,"valid",0.230597475945365,0.939431396786156
    40,"train",0.190593419024368,0.942727647301195
    40,"valid",0.243536252455384,0.936186650185414

With thirty classes to distinguish between, a final validation-set accuracy of \~0.94 looks like a very decent result!

We can confirm this on the test set:

```{r}
evaluate(fitted, test_dl)
```

    loss: 0.2373
    acc: 0.9324

> An interesting question is which words get confused most often. (Of course, even more interesting is how error probabilities are related to features of the spectrograms \-- but this, we have to leave to the *true* domain experts. A nice way of displaying the confusion matrix is to create an alluvial plot. We see the predictions, on the left, "flow into" the target slots. (Target-prediction pairs less frequent than a thousandth of test set cardinality are hidden.)

Reviewer: > the spectrograms \-- but this

Is that slash meant to be there?
